cmd/unpack/main.go

```diff
@@ -24,36 +39,99 @@ func main() {
 				fmt.Printf("Git commit: %s\n", version.String())
 				os.Exit(0)
 			}
-			bundleFS := os.DirFS(bundleDir)
-			bundleContents := map[string][]byte{}
-			if err := fs.WalkDir(bundleFS, ".", func(path string, d fs.DirEntry, err error) error {
-				if err != nil {
-					return err
-				}
-				if d.IsDir() {
-					return nil
-				}
-				data, err := fs.ReadFile(bundleFS, path)
-				if err != nil {
-					return fmt.Errorf("read file %q: %w", path, err)
-				}
-				bundleContents[path] = data
-				return nil
-			}); err != nil {
-				log.Fatalf("walk bundle dir %q: %v", bundleDir, err)
+			contentHandler := serveBundle(os.DirFS(bundleDir))
+			h := mux.NewRouter()
+			h.Handle("/content", contentHandler).Methods(http.MethodGet)
+			srv := http.Server{
+				Addr:         listenAddr,
+				Handler:      handlers.CombinedLoggingHandler(os.Stdout, h),
+				ReadTimeout:  1 * time.Second,
+				WriteTimeout: 10 * time.Second,
 			}
+			go func() {
+				log.Println("listening on", srv.Addr)
+				if err := srv.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+					log.Fatal(err)
+				}
+			}()
 
-			encoder := json.NewEncoder(os.Stdout)
-			if err := encoder.Encode(bundleContents); err != nil {
+			ctx, cancel := signal.NotifyContext(cmd.Context(), syscall.SIGINT, syscall.SIGTERM)
+			defer cancel()
+			<-ctx.Done()
+
+			shutdownCtx, cancelShutdown := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancelShutdown()
+			log.Println("shutting down")
+			if err := srv.Shutdown(shutdownCtx); err != nil {
 				log.Fatal(err)
 			}
+			log.Println("shutdown complete")
+
 			return nil
 		},
 	}
 	cmd.Flags().StringVar(&bundleDir, "bundle-dir", "", "directory in which the bundle can be found")
+	cmd.Flags().StringVar(&listenAddr, "listen-addr", ":8080", "listen address for http server")
 	cmd.Flags().BoolVar(&rukpakVersion, "version", false, "displays rukpak version information")
 
 	if err := cmd.Execute(); err != nil {
 		log.Fatal(err)
 	}
 }
+
+func serveBundle(bundle fs.FS) http.HandlerFunc {
+	return func(writer http.ResponseWriter, request *http.Request) {
+		buf := &bytes.Buffer{}
+		gzw := gzip.NewWriter(buf)
+		tw := tar.NewWriter(gzw)
+		if err := fs.WalkDir(bundle, ".", func(path string, d fs.DirEntry, err error) error {
+			if err != nil {
+				return err
+			}
+
+			if d.Type()&os.ModeSymlink != 0 {
+				return nil
+			}
+
+			info, err := d.Info()
+			if err != nil {
+				return fmt.Errorf("get file info for %q: %w", path, err)
+			}
+
+			h, err := tar.FileInfoHeader(info, "")
+			if err != nil {
+				return fmt.Errorf("build tar file info header for %q: %w", path, err)
+			}
+			h.Uid = 0
+			h.Gid = 0
+			h.Uname = ""
+			h.Gname = ""
+			h.Name = path
```

Comment:
are there any additional file changes that need to be made regarding permissions? I know this is an edge case, but we definitely saw cases in registry where files had strange permissions or x-attrs that causes unpacking to fail on the local filesystem.

Reply:
That's a good question. By the time we're creating this tar, the files are already on the filesystem where the pod is running. I _think_ we're okay, but if not that seems simple enough to fix later?

Reply:
That could be addressed in a follow-up, sure